File: atcoder/go/abc160_169/abc160_F/abc160_F.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const BUFSIZE = 10000000

var rdr = bufio.NewReaderSize(os.Stdin, BUFSIZE)
var wrtr = bufio.NewWriterSize(os.Stdout, BUFSIZE)

func readLine() string {
	buf := make([]byte, 0, 16)
	for {
		l, p, e := rdr.ReadLine()
		if e != nil {
			fmt.Println(e.Error())
			panic(e)
		}
		buf = append(buf, l...)
		if !p {
			break
		}
	}
	return string(buf)
}

func gs() string    { return readLine() }
func gss() []string { return strings.Fields(gs()) }
func gi() int64 {
	res, e := strconv.Atoi(gs())
	if e != nil {
		panic(e)
	}
	return int64(res)
}
func gf() float64 {
	res, e := strconv.ParseFloat(gs(), 64)
	if e != nil {
		panic(e)
	}
	return float64(res)
}
func gis() []int64 {
	res := make([]int64, 0)
	for _, s := range gss() {
		v, e := strconv.Atoi(s)
		if e != nil {
			panic(e)
		}
		res = append(res, int64(v))
	}
	return res
}
func gfs() []float64 {
	res := make([]float64, 0)
	for _, s := range gss() {
		v, _ := strconv.ParseFloat(s, 64)
		res = append(res, float64(v))
	}
	return res
}

const MOD int64 = 1_000_000_007

func powmod(a, e, p int64) int64 {
	var res int64 = 1
	var m int64 = a
	for e > 0 {
		if e&1 != 0 {
			res = res * m % p
		}
		e >>= 1
		m = m * m % p
	}
	return res
}

var sz [200_001]int64
var dp [200_001]int64
var fact [200_001]int64
var factinv [200_001]int64

func dfs(gr [][]int64, n int64, p int64, order []int64) []int64 {
	order = append(order, n)
	// Preprocess all of the children
	for _, c := range gr[n] {
		if c == p {
			continue
		}
		order = dfs(gr, c, n, order)
		order = append(order, n)
	}
	analyze(gr, n, p)
	return order
}

func reroot(gr [][]int64, root int64, nxtroot int64) {
	oldszroot, oldsznxtroot, olddproot, olddpnxtroot := sz[root], sz[nxtroot], dp[root], dp[nxtroot]
	sz[root], sz[nxtroot] = oldszroot-oldsznxtroot, oldszroot
	dp[root] = olddproot * factinv[oldszroot-1] % MOD * fact[sz[root]-1] % MOD * powmod(olddpnxtroot, MOD-2, MOD) % MOD * fact[oldsznxtroot] % MOD
	dp[nxtroot] = olddpnxtroot * factinv[oldsznxtroot-1] % MOD * fact[sz[nxtroot]-1] % MOD * dp[root] % MOD * factinv[sz[root]] % MOD
}

func analyze(gr [][]int64, n int64, p int64) {
	ldp, lsz := int64(1), int64(1)
	for _, c := range gr[n] {
		if c == p {
			continue
		}
		csz := sz[c]
		lsz += csz
		ldp = ldp * dp[c] % MOD * factinv[csz] % MOD
	}
	ldp = ldp * fact[lsz-1] % MOD
	sz[n] = lsz
	dp[n] = ldp
}

func main() {
	//f1, _ := os.Create("cpu.prof")
	//pprof.StartCPUProfile(f1)

	infn := ""
	if infn == "" && len(os.Args) > 1 {
		infn = os.Args[1]
	}
	if infn != "" {
		f, e := os.Open(infn)
		if e != nil {
			panic(e)
		}
		rdr = bufio.NewReaderSize(f, BUFSIZE)
	}
	N := gi()
	gr := make([][]int64, N)
	for i := int64(0); i < N-1; i++ {
		xx := gis()
		a, b := xx[0]-1, xx[1]-1
		gr[a] = append(gr[a], b)
		gr[b] = append(gr[b], a)
	}

	fact[0] = 1
	for i := int64(1); i <= 200_000; i++ {
		fact[i] = fact[i-1] * i % MOD
	}
	factinv[200_000] = powmod(fact[200_000], MOD-2, MOD)
	for i := int64(199_999); i >= 0; i-- {
		factinv[i] = factinv[i+1] * (i + 1) % MOD
	}
	order := make([]int64, 0)
	order = dfs(gr, 0, -1, order)
	ansarr := make([]string, N)
	for i, n := range order {
		ansarr[n] = strconv.FormatInt(dp[n], 10)
		if i+1 < len(order) {
			reroot(gr, n, order[i+1])
		}
	}
	ans := strings.Join(ansarr, "\n")
	fmt.Println(ans)
	//Lots of output -- consolidate into a string
	//anssarr := make([]string, N)
	//for i,a := range ansarr
	//for i := int64(0); i < N; i++ {
	//	fmt.Fprintln(wrtr, ansarr[i])
	//}
	//wrtr.Flush()
	//pprof.StopCPUProfile()
}

File: abc/python/abc160/abc160_F.py

import sys
import collections
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

MOD = 1_000_000_007

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N = gi()
    A = []
    B = []
    for _ in range(N-1) : a,b = gis(); A.append(a-1); B.append(b-1)
    gr = [ [] for _ in range(N) ]
    for (a,b) in zip(A,B) : gr[a].append(b); gr[b].append(a)
    fact = [1] * 200_001
    for i in range(1,200_001) : fact[i] = fact[i-1] * i % MOD
    factinv = [1] * 200_001
    factinv[200_000] = pow(fact[200_000],MOD-2,MOD)
    for i in range(199_999,-1,-1) : factinv[i] = factinv[i+1] * (i+1) % MOD
    sz = [0] * 200_001
    dp = [0] * 200_001

    ## We need a traversal order, so we have to do the poor-mans recursionless
    ## DFS because of python
    order = []
    q = [(0,-1,0)]
    while q :
        (n,p,mode) = q.pop()
        order.append(n)
        if mode == 0 :
            for c in gr[n] :
                if c == p : continue
                q.append((n,p,1))
                q.append((c,n,0))

    ## We can do the DP with just a bottoms up traversal which we can
    ## get from the reverse of a BFS order
    q = collections.deque()
    q.append((0,-1))
    tdorder = []
    while q :
        (n,p) = q.popleft()
        tdorder.append((n,p))
        for c in gr[n] :
            if c == p : continue
            q.append((c,n))
    buorder = tdorder[::-1]

    for (n,p) in buorder :
        lsz,ldp = 1,1
        for c in gr[n] :
            if c == p : continue
            lsz += sz[c]
            ldp = ldp * dp[c] % MOD * factinv[sz[c]] % MOD
        ldp = ldp * fact[lsz-1] % MOD
        sz[n] = lsz
        dp[n] = ldp

    ans = [0] * N
    for i,n in enumerate(order) :
        ans[n] = dp[n]
        if i+1 < len(order) :
            ## Reroot the the tree
            root,nxtroot = n,order[i+1]
            oldszroot,oldsznxtroot = sz[root],sz[nxtroot]
            olddproot,olddpnxtroot = dp[root],dp[nxtroot]
            sz[root], sz[nxtroot] = oldszroot-oldsznxtroot, oldszroot
            dp[root] = olddproot * factinv[oldszroot-1] % MOD * fact[sz[root]-1] % MOD * pow(olddpnxtroot, MOD-2, MOD) % MOD * fact[oldsznxtroot] % MOD
            dp[nxtroot] = olddpnxtroot * factinv[oldsznxtroot-1] % MOD * fact[sz[nxtroot]-1] % MOD * dp[root] % MOD * factinv[sz[root]] % MOD
 
    ansstr = "\n".join([str(x) for x in ans])
    print(ansstr)

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

